test/HLSL/passes/multi_dim_one_dim/gep_addrspacecast_gep.ll

```diff
@@ -0,0 +1,103 @@
+; RUN: opt -S -multi-dim-one-dim %s | FileCheck %s
+;
+; Tests for the pass that changes multi-dimension global variable accesses into
+; a flattened one-dimensional access. The tests focus on the case where the geps
+; need to be merged but are separated by an addrspacecast operation. This was
+; causing the pass to fail because it could not merge the gep through the
+; addrspace cast.
+
+; Naming convention: gep0_addrspacecast_gep1
+
+target datalayout = "e-m:e-p:32:32-i1:32-i8:32-i16:32-i32:32-i64:64-f16:32-f32:32-f64:64-n8:16:32:64"
+target triple = "dxil-ms-dx"
+
+@ArrayOfArray = addrspace(3) global [256 x [9 x float]] undef, align 4
+
+; Test that we can merge the geps when all parts are instructions.
+; CHECK-LABEL: @merge_gep_instr_instr_instr
+; CHECK:  load float, float* addrspacecast (float addrspace(3)* getelementptr inbounds ([2304 x float], [2304 x float] addrspace(3)* @ArrayOfArray.1dim, i32 0, i32 1) to float*)
+define void @merge_gep_instr_instr_instr() {
+entry:
+  %gep0 = getelementptr inbounds [256 x [9 x float]], [256 x [9 x float]] addrspace(3)* @ArrayOfArray, i32 0, i32 0
+  %asc  = addrspacecast [9 x float] addrspace(3)* %gep0 to [9 x float]*
+  %gep1 = getelementptr inbounds [9 x float], [9 x float]* %asc, i32 0, i32 1
```

Comment:
Should we have a test where there are two GEPs after the addrspacecast to show that it successfully migrates the addrspacecast fully across all the GEPs?  It's something I wasn't sure would work properly until I thought more carefully about the algorithm.  I think it will work now, but it would be good to have a test that verifies this.

Reply:
Added a few new tests to cover this case. It uncovered a bug with an early return in the `MergeGEP` function that triggers when looking through the addrspace cast and the GEP ends in an array.

Fixed the bug and added new tests to cover both the addrspacecast case and the "plain" case.